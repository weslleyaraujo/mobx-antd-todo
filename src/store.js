// @flow
import { decorate, observable, action, computed } from "mobx";

type Status = "done" | "";

export type Todo = {
  text: string,
  id: string,
  status: Status
};

type TodosType = Array<Todo>;

type addTodo = { text: string, id: string };
type setStatus = { id: string, status: Status };

export type Props = {
  filter: Status,
  todos: TodosType
};

export type Store = Props & {
  reset: () => void,
  addTodo: addTodo => void,
  setFilter: Status => void,
  setStatus: setStatus => void,
  count: number,
  filtered: TodosType
};

export const Todos = decorate(
  class {
    filter = "";
    todos = [];

    get count(): number {
      return this.todos.length;
    }

    get filtered(): Array<Todo> {
      const { filter } = this;
      if (!Boolean(filter)) {
        return this.todos;
      }

      return this.todos.filter(s => s.status === filter);
    }

    reset(): void {
      this.filter = "";
      this.todos = [];
    }

    addTodo({ text, id }: addTodo) {
      this.todos.push({ id, text, status: "" });
    }

    setFilter(type: Status) {
      this.filter = type;
    }

    setStatus({ id, status }: setStatus) {
      this.todos = this.todos.map(todo => ({
        ...todo,
        status: todo.id === id ? status : todo.status
      }));
    }
  },
  {
    filter: observable,
    todos: observable,
    count: computed,
    filtered: computed,
    reset: action,
    addTodo: action,
    setFilter: action,
    setStatus: action
  }
);

const todos = new Todos();

export default todos;
