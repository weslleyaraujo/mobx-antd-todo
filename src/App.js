// @flow
import React from 'react';
import { observer } from 'mobx-react';
import { Layout, Col, Row, Card, Tag } from 'antd';
import { Box } from 'grid-styled';

import store from './todo-store';
// NOTE: for some reason import type Store leads to a
// "no-unused-vars" warning on eslint.
// eslint-disable-next-line
import type Store from './todo-store';
import Todos from './components/Todos';
import AddTodo from './components/AddTodo';
import './styles.css';
import 'antd/dist/antd.css';

const sleep = (ms: number = 1000): Promise<void> =>
  new Promise(r => setTimeout(r, ms));

function App({ store: Store }) {
  return (
    <Layout>
      <Box p={3}>
        <Row justify="center" span={24} type="flex">
          <Col span={24} lg={{ span: 12 }} xl={{ span: 12 }}>
            <Box px={2}>
              <Card title="Things to be done 🐜">
                {Boolean(store.count) && (
                  <Box py={2}>
                    <Box pb={2}>
                      <Tag.CheckableTag
                        checked={store.filter === ''}
                        onChange={e => store.setFilter('')}
                      >
                        All
                      </Tag.CheckableTag>
                      <Tag.CheckableTag
                        checked={store.filter === 'done'}
                        onChange={e => store.setFilter('done')}
                      >
                        Done
                      </Tag.CheckableTag>
                    </Box>
                    <Todos
                      dataSource={store.filtered}
                      onDoubleClick={todo => {
                        const { id, status } = todo;
                        store.setStatus({
                          id,
                          status: status === 'done' ? '' : 'done'
                        });
                      }}
                    />
                  </Box>
                )}
                <Box py={2}>
                  <AddTodo
                    onSubmit={async (values, { setSubmitting, resetForm }) => {
                      setSubmitting(true);
                      await sleep();
                      store.addTodo({
                        text: values.text,
                        id: String(+new Date())
                      });
                      setSubmitting(false);
                      resetForm();
                    }}
                  />
                </Box>
              </Card>
            </Box>
          </Col>
        </Row>
      </Box>
    </Layout>
  );
}

export default observer(App);
