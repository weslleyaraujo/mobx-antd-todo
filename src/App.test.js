import React from "react";
import store from "./store";
import App from "./App";

it("renders App correctly", () => {
  const wrapper = shallow(<App store={store} />);
  expect(toJson(wrapper)).toMatchSnapshot();
});
