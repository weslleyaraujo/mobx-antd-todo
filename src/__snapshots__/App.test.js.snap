// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders App correctly 1`] = `
<Adapter>
  <Box
    p={3}
    theme={
      Object {
        "breakpoints": Array [
          "40em",
          "52em",
          "64em",
        ],
        "fontSizes": Array [
          12,
          14,
          16,
          20,
          24,
          32,
          48,
          64,
          96,
          128,
        ],
        "space": Array [
          0,
          4,
          8,
          16,
          32,
          64,
          128,
          256,
          512,
        ],
      }
    }
  >
    <Row
      gutter={0}
      justify="center"
      span={24}
      type="flex"
    >
      <Col
        lg={
          Object {
            "span": 12,
          }
        }
        span={24}
        xl={
          Object {
            "span": 12,
          }
        }
      >
        <Box
          px={2}
          theme={
            Object {
              "breakpoints": Array [
                "40em",
                "52em",
                "64em",
              ],
              "fontSizes": Array [
                12,
                14,
                16,
                20,
                24,
                32,
                48,
                64,
                96,
                128,
              ],
              "space": Array [
                0,
                4,
                8,
                16,
                32,
                64,
                128,
                256,
                512,
              ],
            }
          }
        >
          <Card
            title="Things to be done 🐜"
          >
            <Box
              py={2}
              theme={
                Object {
                  "breakpoints": Array [
                    "40em",
                    "52em",
                    "64em",
                  ],
                  "fontSizes": Array [
                    12,
                    14,
                    16,
                    20,
                    24,
                    32,
                    48,
                    64,
                    96,
                    128,
                  ],
                  "space": Array [
                    0,
                    4,
                    8,
                    16,
                    32,
                    64,
                    128,
                    256,
                    512,
                  ],
                }
              }
            >
              <AddTodo
                onSubmit={[Function]}
              />
            </Box>
          </Card>
        </Box>
      </Col>
    </Row>
  </Box>
</Adapter>
`;
